Index: src/minecraft/net/minecraft/block/Block.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/block/Block.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/block/Block.java	(date 1502999984000)
@@ -2,6 +2,9 @@
 
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
+
+import de.paxii.clarinet.module.render.ModuleXray;
+
 import java.util.List;
 import java.util.Random;
 import java.util.Set;
@@ -506,6 +509,13 @@
     @Deprecated
     public boolean shouldSideBeRendered(IBlockState blockState, IBlockAccess blockAccess, BlockPos pos, EnumFacing side)
     {
+        //TODO: Client
+            if (ModuleXray.isActive()) {
+                if (ModuleXray.getBlockList().contains(REGISTRY.getIDForObject(this))) {
+                    return true;
+                }
+            }
+        //
         AxisAlignedBB axisalignedbb = blockState.getBoundingBox(blockAccess, pos);
 
         switch (side)
@@ -828,6 +838,16 @@
 
     public BlockRenderLayer getBlockLayer()
     {
+        //TODO: Client
+            if (ModuleXray.isActive()) {
+                if (ModuleXray.getBlockList().contains(REGISTRY.getIDForObject(this))) {
+                    return BlockRenderLayer.SOLID;
+                }
+
+                return BlockRenderLayer.TRANSLUCENT;
+            }
+        //
+
         return BlockRenderLayer.SOLID;
     }
 
@@ -1526,4 +1546,10 @@
         XZ,
         XYZ;
     }
+
+    //TODO: Client
+        public Material getMaterial() {
+            return this.blockMaterial;
+        }
+    //
 }
Index: src/minecraft/net/minecraft/block/BlockBush.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/block/BlockBush.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/block/BlockBush.java	(date 1502999984000)
@@ -1,5 +1,7 @@
 package net.minecraft.block;
 
+import de.paxii.clarinet.module.render.ModuleXray;
+
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.MapColor;
@@ -105,6 +107,12 @@
 
     public BlockRenderLayer getBlockLayer()
     {
+        //TODO: Client
+            if (ModuleXray.isActive()) {
+                return super.getBlockLayer();
+            }
+        //
+
         return BlockRenderLayer.CUTOUT;
     }
 
Index: src/minecraft/net/minecraft/block/BlockCactus.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/block/BlockCactus.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/block/BlockCactus.java	(date 1502999984000)
@@ -1,5 +1,7 @@
 package net.minecraft.block;
 
+import de.paxii.clarinet.module.render.ModuleXray;
+
 import java.util.Random;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.properties.IProperty;
@@ -134,6 +136,12 @@
 
     public BlockRenderLayer getBlockLayer()
     {
+        //TODO: Client
+            if (ModuleXray.isActive()) {
+                return super.getBlockLayer();
+            }
+        //
+
         return BlockRenderLayer.CUTOUT;
     }
 
Index: src/minecraft/net/minecraft/block/BlockGrass.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/block/BlockGrass.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/block/BlockGrass.java	(date 1502999984000)
@@ -1,5 +1,7 @@
 package net.minecraft.block;
 
+import de.paxii.clarinet.module.render.ModuleXray;
+
 import java.util.Random;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.properties.IProperty;
@@ -145,6 +147,12 @@
 
     public BlockRenderLayer getBlockLayer()
     {
+        //TODO: Client
+            if (ModuleXray.isActive()) {
+                return super.getBlockLayer();
+            }
+        //
+
         return BlockRenderLayer.CUTOUT_MIPPED;
     }
 
Index: src/minecraft/net/minecraft/block/BlockLeaves.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/block/BlockLeaves.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/block/BlockLeaves.java	(date 1502999984000)
@@ -1,5 +1,7 @@
 package net.minecraft.block;
 
+import de.paxii.clarinet.module.render.ModuleXray;
+
 import java.util.Random;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.properties.PropertyBool;
@@ -277,6 +279,12 @@
 
     public BlockRenderLayer getBlockLayer()
     {
+        //TODO: Client
+            if (ModuleXray.isActive()) {
+                return super.getBlockLayer();
+            }
+        //
+
         return this.leavesFancy ? BlockRenderLayer.CUTOUT_MIPPED : BlockRenderLayer.SOLID;
     }
 
Index: src/minecraft/net/minecraft/client/Minecraft.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/Minecraft.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/Minecraft.java	(date 1502999984000)
@@ -13,6 +13,13 @@
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.authlib.properties.PropertyMap;
 import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
+
+import de.paxii.clarinet.Client;
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.game.StartGameEvent;
+import de.paxii.clarinet.event.events.gui.DisplayGuiScreenEvent;
+import de.paxii.clarinet.util.protocol.ProtocolVersion;
+
 import java.awt.image.BufferedImage;
 import java.io.File;
 import java.io.IOException;
@@ -218,6 +225,9 @@
 import org.lwjgl.opengl.PixelFormat;
 import org.lwjgl.util.glu.GLU;
 
+import lombok.Getter;
+import lombok.Setter;
+
 public class Minecraft implements IThreadListener, ISnooperInfo
 {
     private static final Logger LOGGER = LogManager.getLogger();
@@ -252,7 +262,11 @@
 
     /** True if the player is connected to a realms server */
     private boolean connectedToRealms;
-    private final Timer timer = new Timer(20.0F);
+    //TODO: Client
+        @Getter
+        @Setter
+    //
+    private Timer timer = new Timer(20.0F);
 
     /** Instance of PlayerUsageSnooper. */
     private final Snooper usageSnooper = new Snooper("client", this, MinecraftServer.getCurrentTimeMillis());
@@ -269,7 +283,10 @@
 
     /** Manages all search trees */
     private SearchTreeManager searchTreeManager = new SearchTreeManager();
-    private final Session session;
+    //TODO: Client
+        @Setter
+        private Session session;
+    //
     private boolean isGamePaused;
 
     /**
@@ -301,7 +318,11 @@
 
     /** Instance of IntegratedServer. */
     private IntegratedServer integratedServer;
-    public GuiIngame ingameGUI;
+
+    //TODO: Client
+        @Getter
+        public GuiIngame ingameGUI;
+    //
 
     /** Skip render world */
     public boolean skipRenderWorld;
@@ -331,7 +352,11 @@
     /**
      * When you place a block, it's set to 6, decremented once per tick, when it's 0, you can place another block.
      */
-    private int rightClickDelayTimer;
+    //TODO: Client
+        @Getter
+        @Setter
+        private int rightClickDelayTimer;
+    //
     private String serverName;
     private int serverPort;
 
@@ -608,6 +633,10 @@
         this.checkGLError("Post startup");
         this.ingameGUI = new GuiIngame(this);
 
+        //TODO: Client
+            new Client();
+        //
+
         if (this.serverName != null)
         {
             this.displayGuiScreen(new GuiConnecting(new GuiMainMenu(), this, this.serverName, this.serverPort));
@@ -638,6 +667,10 @@
         }
 
         this.renderGlobal.makeEntityOutlineShader();
+
+        //TODO: Client
+            EventManager.call(new StartGameEvent());
+        //
     }
 
     /**
@@ -1060,6 +1093,12 @@
      */
     public void displayGuiScreen(@Nullable GuiScreen guiScreenIn)
     {
+        //TODO: Client
+            DisplayGuiScreenEvent guiScreenEvent = new DisplayGuiScreenEvent(guiScreenIn);
+            EventManager.call(guiScreenEvent);
+            guiScreenIn = guiScreenEvent.getGuiScreen();
+        //
+
         if (this.currentScreen != null)
         {
             this.currentScreen.onGuiClosed();
@@ -3397,7 +3436,12 @@
                     }
                     else if (i == this.gameSettings.keyBindScreenshot.getKeyCode())
                     {
-                        this.ingameGUI.getChatGUI().printChatMessage(ScreenShotHelper.saveScreenshot(this.mcDataDir, this.displayWidth, this.displayHeight, this.framebufferMc));
+                        //TODO: Client
+                            //Fix Screenshot bug when pressing '<' on german keyboards in chat
+                            if (Keyboard.getEventKey() != 0) {
+                                this.ingameGUI.getChatGUI().printChatMessage(ScreenShotHelper.saveScreenshot(this.mcDataDir, this.displayWidth, this.displayHeight, this.framebufferMc));
+                            }
+                        //
                     }
                     else if (i == 48 && GuiScreen.isCtrlKeyDown() && (this.currentScreen == null || this.currentScreen != null && !this.currentScreen.isFocused()))
                     {
Index: src/minecraft/net/minecraft/client/audio/SoundHandler.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/audio/SoundHandler.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/audio/SoundHandler.java	(date 1502999984000)
@@ -30,6 +30,8 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import lombok.Getter;
+
 public class SoundHandler implements IResourceManagerReloadListener, ITickable
 {
     public static final Sound MISSING_SOUND = new Sound("meta:missing_sound", 1.0F, 1.0F, 1, Sound.Type.FILE, false);
@@ -51,9 +53,14 @@
         }
     };
     private final SoundRegistry soundRegistry = new SoundRegistry();
-    private final SoundManager sndManager;
+
     private final IResourceManager mcResourceManager;
 
+    //TODO: Client
+        @Getter
+        private final SoundManager sndManager;
+    //
+
     public SoundHandler(IResourceManager manager, GameSettings gameSettingsIn)
     {
         this.mcResourceManager = manager;
Index: src/minecraft/net/minecraft/client/entity/AbstractClientPlayer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/entity/AbstractClientPlayer.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/entity/AbstractClientPlayer.java	(date 1502999984000)
@@ -1,6 +1,9 @@
 package net.minecraft.client.entity;
 
 import com.mojang.authlib.GameProfile;
+
+import de.paxii.clarinet.util.player.capesapi.CapesApi;
+
 import java.io.File;
 import javax.annotation.Nullable;
 import net.minecraft.client.Minecraft;
@@ -43,6 +46,9 @@
             this.nameClear = StringUtils.stripControlCodes(this.nameClear);
         }
 
+        // TODO: Client
+            CapesApi.loadCape(this.getGameProfile());
+        //
         CapeUtils.downloadCape(this);
         PlayerConfigurations.getPlayerConfiguration(this);
     }
@@ -106,6 +112,12 @@
         {
             return null;
         }
+        // TODO: Client
+            else if (CapesApi.hasCape(this.getGameProfile()))
+            {
+                return CapesApi.getCape(this.getGameProfile());
+            }
+        //
         else if (this.locationOfCape != null)
         {
             return this.locationOfCape;
Index: src/minecraft/net/minecraft/client/entity/EntityPlayerSP.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/entity/EntityPlayerSP.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/entity/EntityPlayerSP.java	(date 1502999984000)
@@ -1,5 +1,12 @@
 package net.minecraft.client.entity;
 
+import de.paxii.clarinet.Wrapper;
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.player.PlayerMoveEvent;
+import de.paxii.clarinet.event.events.player.PlayerUpdateWalkingEvent;
+import de.paxii.clarinet.event.events.player.PostMotionUpdateEvent;
+import de.paxii.clarinet.event.events.player.PreMotionUpdateEvent;
+
 import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.block.state.IBlockState;
@@ -232,10 +239,37 @@
      */
     public void onUpdate()
     {
+        //TODO: Client
+            if (this.motionX != 0 || this.motionY != -0.0784000015258789 || this.motionZ != 0) {
+                PlayerMoveEvent moveEvent = new PlayerMoveEvent(this);
+                EventManager.call(moveEvent);
+
+                if (moveEvent.isCancelled())
+                    return;
+
+                this.motionX = moveEvent.getMotionX();
+                this.motionY = moveEvent.getMotionY();
+                this.motionZ = moveEvent.getMotionZ();
+
+                this.posX = moveEvent.getPosX();
+                this.posY = moveEvent.getPosY();
+                this.posZ = moveEvent.getPosZ();
+            }
+        //
+
         if (this.world.isBlockLoaded(new BlockPos(this.posX, 0.0D, this.posZ)))
         {
             super.onUpdate();
 
+            //TODO: Client
+                PreMotionUpdateEvent preMotionEvent = new PreMotionUpdateEvent(this);
+                EventManager.call(preMotionEvent);
+
+                if (preMotionEvent.isCancelled()) {
+                    return;
+                }
+            //
+
             if (this.isRiding())
             {
                 this.connection.sendPacket(new CPacketPlayer.Rotation(this.rotationYaw, this.rotationPitch, this.onGround));
@@ -251,6 +285,11 @@
             {
                 this.onUpdateWalkingPlayer();
             }
+
+            //TODO: Client
+                PostMotionUpdateEvent postMotion = new PostMotionUpdateEvent(this);
+                EventManager.call(postMotion);
+            //
         }
     }
 
@@ -303,6 +342,12 @@
             boolean flag2 = d0 * d0 + d1 * d1 + d2 * d2 > 9.0E-4D || this.positionUpdateTicks >= 20;
             boolean flag3 = d3 != 0.0D || d4 != 0.0D;
 
+            // TODO: Client
+                PlayerUpdateWalkingEvent updateWalkingEvent = EventManager.call(new PlayerUpdateWalkingEvent());
+
+                if (!updateWalkingEvent.isCancelled())
+            //
+
             if (this.isRiding())
             {
                 this.connection.sendPacket(new CPacketPlayer.PositionRotation(this.motionX, -999.0D, this.motionZ, this.rotationYaw, this.rotationPitch, this.onGround));
@@ -973,9 +1018,13 @@
 
         if (this.isHandActive() && !this.isRiding())
         {
-            this.movementInput.moveStrafe *= 0.2F;
-            this.movementInput.moveForward *= 0.2F;
-            this.sprintToggleTimer = 0;
+            //TODO: Client
+                if (!Wrapper.getModuleManager().isModuleActive("NoItemSlowdown")) {
+                    this.movementInput.moveStrafe *= 0.2F;
+                    this.movementInput.moveForward *= 0.2F;
+                    this.sprintToggleTimer = 0;
+                }
+            //
         }
 
         boolean flag3 = false;
Index: src/minecraft/net/minecraft/client/gui/FontRenderer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/gui/FontRenderer.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/gui/FontRenderer.java	(date 1502999984000)
@@ -29,6 +29,8 @@
 import net.minecraft.util.ResourceLocation;
 import org.apache.commons.io.IOUtils;
 
+import lombok.Getter;
+
 public class FontRenderer implements IResourceManagerReloadListener
 {
     private static final ResourceLocation[] UNICODE_PAGE_LOCATIONS = new ResourceLocation[256];
@@ -49,7 +51,10 @@
      * Array of RGB triplets defining the 16 standard chat colors followed by 16 darker version of the same colors for
      * drop shadows.
      */
-    private final int[] colorCode = new int[32];
+    //TODO: Client
+        @Getter
+        private final int[] colorCode = new int[32];
+    //
     private ResourceLocation locationFontTexture;
 
     /** The RenderEngine used to load and setup glyph textures. */
Index: src/minecraft/net/minecraft/client/gui/Gui.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/gui/Gui.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/gui/Gui.java	(date 1502999984000)
@@ -17,7 +17,9 @@
     /**
      * Draws a thin horizontal line between two points.
      */
-    protected void drawHorizontalLine(int startX, int endX, int y, int color)
+    // TODO: Client
+        protected static void drawHorizontalLine(int startX, int endX, int y, int color)
+    //
     {
         if (endX < startX)
         {
@@ -32,7 +34,9 @@
     /**
      * Draw a 1 pixel wide vertical line. Args : x, y1, y2, color
      */
-    protected void drawVerticalLine(int x, int startY, int endY, int color)
+    // TODO: Client
+        protected static void drawVerticalLine(int x, int startY, int endY, int color)
+    //
     {
         if (endY < startY)
         {
@@ -79,6 +83,9 @@
         bufferbuilder.pos((double)right, (double)top, 0.0D).endVertex();
         bufferbuilder.pos((double)left, (double)top, 0.0D).endVertex();
         tessellator.draw();
+        //TODO: Client
+            GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+        //
         GlStateManager.enableTexture2D();
         GlStateManager.disableBlend();
     }
Index: src/minecraft/net/minecraft/client/gui/GuiChat.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/gui/GuiChat.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/gui/GuiChat.java	(date 1502999984000)
@@ -15,6 +15,9 @@
 import org.lwjgl.input.Keyboard;
 import org.lwjgl.input.Mouse;
 
+import lombok.Getter;
+import lombok.Setter;
+
 public class GuiChat extends GuiScreen implements ITabCompleter
 {
     private static final Logger LOGGER = LogManager.getLogger();
@@ -33,7 +36,11 @@
     /**
      * is the text that appears when you press the chat key and the input box appears pre-filled
      */
-    private String defaultInputFieldText = "";
+    //TODO: Client
+        @Getter
+        @Setter
+        private String defaultInputFieldText = "";
+    //
 
     public GuiChat()
     {
Index: src/minecraft/net/minecraft/client/gui/GuiScreen.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/gui/GuiScreen.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/gui/GuiScreen.java	(date 1502999984000)
@@ -3,6 +3,10 @@
 import com.google.common.base.Splitter;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
+
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.gui.RenderGuiScreenEvent;
+
 import java.awt.Toolkit;
 import java.awt.datatransfer.ClipboardOwner;
 import java.awt.datatransfer.DataFlavor;
@@ -82,6 +86,11 @@
      */
     public void drawScreen(int mouseX, int mouseY, float partialTicks)
     {
+        //TODO: Client
+            RenderGuiScreenEvent renderScreen = new RenderGuiScreenEvent(mouseX, mouseY, partialTicks);
+            EventManager.call(renderScreen);
+        //
+
         for (int i = 0; i < this.buttonList.size(); ++i)
         {
             ((GuiButton)this.buttonList.get(i)).drawButton(this.mc, mouseX, mouseY, partialTicks);
Index: src/minecraft/net/minecraft/client/gui/GuiSlot.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/gui/GuiSlot.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/gui/GuiSlot.java	(date 1502999984000)
@@ -23,7 +23,9 @@
     protected int left;
 
     /** The height of a slot. */
-    protected final int slotHeight;
+    //TODO: Client
+        protected int slotHeight;
+    //
 
     /** The buttonID of the button used to scroll up */
     private int scrollUpButtonID;
Index: src/minecraft/net/minecraft/client/gui/GuiTextField.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/gui/GuiTextField.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/gui/GuiTextField.java	(date 1502999984000)
@@ -10,53 +10,56 @@
 import net.minecraft.util.ChatAllowedCharacters;
 import net.minecraft.util.math.MathHelper;
 
+/**
+ * TODO: Client
+ */
 public class GuiTextField extends Gui
 {
-    private final int id;
-    private final FontRenderer fontRenderer;
+    protected final int id;
+    protected final FontRenderer fontRenderer;
     public int x;
     public int y;
 
     /** The width of this text field. */
-    private final int width;
-    private final int height;
+    protected final int width;
+    protected final int height;
 
     /** Has the current text being edited on the textbox. */
-    private String text = "";
-    private int maxStringLength = 32;
-    private int cursorCounter;
-    private boolean enableBackgroundDrawing = true;
+    protected String text = "";
+    protected int maxStringLength = 32;
+    protected int cursorCounter;
+    protected boolean enableBackgroundDrawing = true;
 
     /**
      * if true the textbox can lose focus by clicking elsewhere on the screen
      */
-    private boolean canLoseFocus = true;
+    protected boolean canLoseFocus = true;
 
     /**
      * If this value is true along with isEnabled, keyTyped will process the keys.
      */
-    private boolean isFocused;
+    protected boolean isFocused;
 
     /**
      * If this value is true along with isFocused, keyTyped will process the keys.
      */
-    private boolean isEnabled = true;
+    protected boolean isEnabled = true;
 
     /**
      * The current character index that should be used as start of the rendered text.
      */
-    private int lineScrollOffset;
-    private int cursorPosition;
+    protected int lineScrollOffset;
+    protected int cursorPosition;
 
     /** other selection position, maybe the same as the cursor */
-    private int selectionEnd;
-    private int enabledColor = 14737632;
-    private int disabledColor = 7368816;
+    protected int selectionEnd;
+    protected int enabledColor = 14737632;
+    protected int disabledColor = 7368816;
 
     /** True if this textbox is visible */
-    private boolean visible = true;
-    private GuiPageButtonList.GuiResponder guiResponder;
-    private Predicate<String> validator = Predicates.<String>alwaysTrue();
+    protected boolean visible = true;
+    protected GuiPageButtonList.GuiResponder guiResponder;
+    protected Predicate<String> validator = Predicates.<String>alwaysTrue();
 
     public GuiTextField(int componentId, FontRenderer fontrendererObj, int x, int y, int par5Width, int par6Height)
     {
@@ -612,7 +615,7 @@
     /**
      * Draws the blue selection box.
      */
-    private void drawSelectionBox(int startX, int startY, int endX, int endY)
+    protected void drawSelectionBox(int startX, int startY, int endX, int endY)
     {
         if (startX < endX)
         {
Index: src/minecraft/net/minecraft/client/gui/ServerListEntryNormal.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/gui/ServerListEntryNormal.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/gui/ServerListEntryNormal.java	(date 1502999984000)
@@ -1,6 +1,9 @@
 package net.minecraft.client.gui;
 
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+import de.paxii.clarinet.util.protocol.ProtocolVersion;
+
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufInputStream;
 import io.netty.buffer.Unpooled;
@@ -76,8 +79,10 @@
             });
         }
 
-        boolean flag = this.server.version > 338;
-        boolean flag1 = this.server.version < 338;
+        // TODO: Client
+            boolean flag = this.server.version > ProtocolVersion.getProtocolVersion();
+            boolean flag1 = this.server.version < ProtocolVersion.getProtocolVersion();
+        //
         boolean flag2 = flag || flag1;
         this.mc.fontRenderer.drawString(this.server.serverName, x + 32 + 3, y + 1, 16777215);
         List<String> list = this.mc.fontRenderer.listFormattedStringToWidth(this.server.serverMOTD, listWidth - 32 - 2);
Index: src/minecraft/net/minecraft/client/gui/inventory/GuiEditSign.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/gui/inventory/GuiEditSign.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/gui/inventory/GuiEditSign.java	(date 1502999984000)
@@ -1,5 +1,9 @@
 package net.minecraft.client.gui.inventory;
 
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.gui.sign.StartEditSignEvent;
+import de.paxii.clarinet.event.events.gui.sign.StopEditSignEvent;
+
 import java.io.IOException;
 import net.minecraft.block.Block;
 import net.minecraft.client.gui.GuiButton;
@@ -44,6 +48,22 @@
         Keyboard.enableRepeatEvents(true);
         this.doneBtn = this.addButton(new GuiButton(0, this.width / 2 - 100, this.height / 4 + 120, I18n.format("gui.done")));
         this.tileSign.setEditable(false);
+
+        // TODO: Client
+            StartEditSignEvent startEditSignEvent = new StartEditSignEvent(this.tileSign);
+            EventManager.call(startEditSignEvent);
+
+            for (int i = 0; i < startEditSignEvent.getSignText().length; i++) {
+                this.tileSign.signText[i] = startEditSignEvent.getSignText()[i];
+            }
+
+            this.editLine = startEditSignEvent.getEditLine();
+
+            if (startEditSignEvent.isCloseGui()) {
+                this.tileSign.markDirty();
+                this.mc.displayGuiScreen(null);
+            }
+        //
     }
 
     /**
@@ -51,6 +71,17 @@
      */
     public void onGuiClosed()
     {
+        // TODO: Client
+            StopEditSignEvent stopEditSignEvent = new StopEditSignEvent(this.tileSign);
+            EventManager.call(stopEditSignEvent);
+
+            for (int i = 0; i < stopEditSignEvent.getSignText().length; i++) {
+                this.tileSign.signText[i] = stopEditSignEvent.getSignText()[i];
+            }
+
+            this.editLine = stopEditSignEvent.getEditLine();
+        //
+
         Keyboard.enableRepeatEvents(false);
         NetHandlerPlayClient nethandlerplayclient = this.mc.getConnection();
 
Index: src/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java	(date 1502999984000)
@@ -1,5 +1,11 @@
 package net.minecraft.client.multiplayer;
 
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.block.BlockBreakEvent;
+import de.paxii.clarinet.event.events.player.PlayerAttackEntityEvent;
+import de.paxii.clarinet.event.events.player.PlayerClickBlockEvent;
+import de.paxii.clarinet.event.events.player.PostPlayerAttackEntityEvent;
+
 import io.netty.buffer.Unpooled;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockCommandBlock;
@@ -174,6 +180,11 @@
             }
             else
             {
+                //TODO: Client
+                    BlockBreakEvent blockBreakEvent = new BlockBreakEvent(block, pos);
+                    EventManager.call(blockBreakEvent);
+                //
+
                 world.playEvent(2001, pos, Block.getStateId(iblockstate));
                 block.onBlockHarvested(world, pos, iblockstate, this.mc.player);
                 boolean flag = world.setBlockState(pos, Blocks.AIR.getDefaultState(), 11);
@@ -210,6 +221,15 @@
      */
     public boolean clickBlock(BlockPos loc, EnumFacing face)
     {
+        //TODO: Client
+            PlayerClickBlockEvent clickBlockEvent = new PlayerClickBlockEvent(loc, face);
+            EventManager.call(clickBlockEvent);
+
+            if (clickBlockEvent.isCancelled()) {
+                return false;
+            }
+        //
+
         if (this.currentGameType.isAdventure())
         {
             if (this.currentGameType == GameType.SPECTATOR)
@@ -530,6 +550,17 @@
      */
     public void attackEntity(EntityPlayer playerIn, Entity targetEntity)
     {
+        //TODO: Client
+            PlayerAttackEntityEvent attackEvent = new PlayerAttackEntityEvent(playerIn, targetEntity);
+            EventManager.call(attackEvent);
+
+            if (attackEvent.isCancelled()) {
+                return;
+            }
+
+            targetEntity = attackEvent.getTarget();
+        //
+
         this.syncCurrentPlayItem();
         this.connection.sendPacket(new CPacketUseEntity(targetEntity));
 
@@ -538,6 +569,11 @@
             playerIn.attackTargetEntityWithCurrentItem(targetEntity);
             playerIn.resetCooldown();
         }
+
+        //TODO: Client
+            PostPlayerAttackEntityEvent postAttackEvent = new PostPlayerAttackEntityEvent(playerIn, targetEntity);
+            EventManager.call(postAttackEvent);
+        //
     }
 
     /**
Index: src/minecraft/net/minecraft/client/multiplayer/ServerData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/multiplayer/ServerData.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/multiplayer/ServerData.java	(date 1502999984000)
@@ -1,5 +1,7 @@
 package net.minecraft.client.multiplayer;
 
+import de.paxii.clarinet.util.protocol.ProtocolVersion;
+
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.TextComponentTranslation;
@@ -23,7 +25,9 @@
 
     /** last server ping that showed up in the server browser */
     public long pingToServer;
-    public int version = 338;
+    // TODO: Client
+        public int version = ProtocolVersion.getProtocolVersion();
+    //
 
     /** Game version for this server. */
     public String gameVersion = "1.12.1";
Index: src/minecraft/net/minecraft/client/multiplayer/WorldClient.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/multiplayer/WorldClient.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/multiplayer/WorldClient.java	(date 1502999984000)
@@ -1,6 +1,10 @@
 package net.minecraft.client.multiplayer;
 
 import com.google.common.collect.Sets;
+
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.game.QuitServerEvent;
+
 import java.util.Random;
 import java.util.Set;
 import javax.annotation.Nullable;
@@ -349,6 +353,10 @@
      */
     public void sendQuittingDisconnectingPacket()
     {
+        //TODO: Client
+            EventManager.call(new QuitServerEvent());
+        //
+
         this.connection.getNetworkManager().closeChannel(new TextComponentString("Quitting"));
     }
 
Index: src/minecraft/net/minecraft/client/network/NetHandlerPlayClient.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/network/NetHandlerPlayClient.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/network/NetHandlerPlayClient.java	(date 1502999984000)
@@ -4,6 +4,12 @@
 import com.google.common.util.concurrent.FutureCallback;
 import com.google.common.util.concurrent.Futures;
 import com.mojang.authlib.GameProfile;
+
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.PacketPatcher;
+import de.paxii.clarinet.event.events.chat.ReceiveChatEvent;
+import de.paxii.clarinet.event.events.entity.EntityVelocityEvent;
+
 import io.netty.buffer.Unpooled;
 import java.io.File;
 import java.io.IOException;
@@ -574,6 +580,17 @@
      */
     public void handleEntityVelocity(SPacketEntityVelocity packetIn)
     {
+        //TODO: Client
+            EntityVelocityEvent velocityEvent = new EntityVelocityEvent(packetIn);
+            EventManager.call(velocityEvent);
+
+            if (velocityEvent.isCancelled()) {
+                return;
+            }
+
+            packetIn = velocityEvent.getVelocityPacket();
+        //
+
         PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.gameController);
         Entity entity = this.clientWorldController.getEntityByID(packetIn.getEntityID());
 
@@ -893,9 +910,23 @@
 
     public void sendPacket(Packet<?> packetIn)
     {
+        //TODO: Client
+            packetIn = PacketPatcher.getPatchedPacket(packetIn);
+
+            if (packetIn == null) {
+                return;
+            }
+        //
+
         this.netManager.sendPacket(packetIn);
     }
 
+    //TODO: Client
+        public void addToSendQueue(Packet<?> packetIn) {
+            this.sendPacket(packetIn);
+        }
+    //
+
     public void handleCollectItem(SPacketCollectItem packetIn)
     {
         PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.gameController);
@@ -933,6 +964,20 @@
      */
     public void handleChat(SPacketChat packetIn)
     {
+        //TODO: Client
+            ReceiveChatEvent chatEvent = new ReceiveChatEvent(packetIn.getChatComponent().getFormattedText());
+
+            EventManager.call(chatEvent);
+
+            if (chatEvent.isCancelled()) {
+                return;
+            }
+
+            if (!chatEvent.getChatMessage().equals(packetIn.getChatComponent().getFormattedText())) {
+                packetIn = new SPacketChat(new TextComponentTranslation(chatEvent.getChatMessage()), packetIn.getType());
+            }
+        //
+
         PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.gameController);
         this.gameController.ingameGUI.addChatMessage(packetIn.getType(), packetIn.getChatComponent());
     }
Index: src/minecraft/net/minecraft/client/renderer/BufferBuilder.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/renderer/BufferBuilder.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/renderer/BufferBuilder.java	(date 1502999984000)
@@ -1,6 +1,9 @@
 package net.minecraft.client.renderer;
 
 import com.google.common.primitives.Floats;
+
+import de.paxii.clarinet.module.render.ModuleXray;
+
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
 import java.nio.FloatBuffer;
@@ -430,27 +433,33 @@
         int i = this.getColorIndex(vertexIndex);
         int j = -1;
 
-        if (!this.noColor)
-        {
-            j = this.rawIntBuffer.get(i);
+        // TODO: Client
+            if (ModuleXray.isActive()) {
+                j = ModuleXray.getXrayOpacity();
+            } else {
+                if (!this.noColor)
+                {
+                    j = this.rawIntBuffer.get(i);
 
-            if (ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN)
-            {
-                int k = (int)((float)(j & 255) * red);
-                int l = (int)((float)(j >> 8 & 255) * green);
-                int i1 = (int)((float)(j >> 16 & 255) * blue);
-                j = j & -16777216;
-                j = j | i1 << 16 | l << 8 | k;
-            }
-            else
-            {
-                int j1 = (int)((float)(j >> 24 & 255) * red);
-                int k1 = (int)((float)(j >> 16 & 255) * green);
-                int l1 = (int)((float)(j >> 8 & 255) * blue);
-                j = j & 255;
-                j = j | j1 << 24 | k1 << 16 | l1 << 8;
-            }
-        }
+                    if (ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN)
+                    {
+                        int k = (int)((float)(j & 255) * red);
+                        int l = (int)((float)(j >> 8 & 255) * green);
+                        int i1 = (int)((float)(j >> 16 & 255) * blue);
+                        j = j & -16777216;
+                        j = j | i1 << 16 | l << 8 | k;
+                    }
+                    else
+                    {
+                        int j1 = (int)((float)(j >> 24 & 255) * red);
+                        int k1 = (int)((float)(j >> 16 & 255) * green);
+                        int l1 = (int)((float)(j >> 8 & 255) * blue);
+                        j = j & 255;
+                        j = j | j1 << 24 | k1 << 16 | l1 << 8;
+                    }
+                }
+            }
+        //
 
         this.rawIntBuffer.put(i, j);
     }
Index: src/minecraft/net/minecraft/client/renderer/EntityRenderer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/renderer/EntityRenderer.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/renderer/EntityRenderer.java	(date 1502999984000)
@@ -3,6 +3,11 @@
 import com.google.common.base.Predicate;
 import com.google.common.base.Predicates;
 import com.google.gson.JsonSyntaxException;
+
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.game.IngameTickEvent;
+import de.paxii.clarinet.event.events.game.RenderTickEvent;
+
 import java.awt.Graphics;
 import java.awt.image.BufferedImage;
 import java.awt.image.ImageObserver;
@@ -95,6 +100,8 @@
 import org.lwjgl.opengl.GLContext;
 import org.lwjgl.util.glu.GLU;
 import org.lwjgl.util.glu.Project;
+
+import lombok.Getter;
 import shadersmod.client.Shaders;
 import shadersmod.client.ShadersRender;
 
@@ -120,7 +127,10 @@
     private int rendererUpdateCount;
 
     /** Pointed entity */
-    private Entity pointedEntity;
+    //TODO: Client
+        @Getter
+        private Entity pointedEntity;
+    //
     private MouseFilter mouseFilterXAxis = new MouseFilter();
     private MouseFilter mouseFilterYAxis = new MouseFilter();
     private final float thirdPersonDistance = 4.0F;
@@ -1378,6 +1388,10 @@
                     }
                 }
 
+                //TODO: Client
+                    EventManager.call(new IngameTickEvent());
+                //
+
                 this.mc.mcProfiler.endSection();
             }
 
@@ -1769,6 +1783,10 @@
             GlStateManager.setFogEnabled(flag2);
         }
 
+        //TODO: Client
+            EventManager.call(new RenderTickEvent(partialTicks));
+        //
+
         if (!renderglobal.damagedBlocks.isEmpty())
         {
             this.mc.mcProfiler.endStartSection("destroyProgress");
Index: src/minecraft/net/minecraft/client/renderer/RenderGlobal.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/renderer/RenderGlobal.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/renderer/RenderGlobal.java	(date 1502999984000)
@@ -5,6 +5,8 @@
 import com.google.common.collect.Sets;
 import com.google.gson.JsonSyntaxException;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+
+import java.awt.*;
 import java.io.IOException;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
@@ -3589,4 +3591,41 @@
             this.setFacing = p_initialize_3_;
         }
     }
+
+    //TODO: Client
+        public static void drawOutlinedBoundingBox(AxisAlignedBB axisAlignedBB, int color) {
+            Color decode = new Color(color);
+            RenderGlobal.drawOutlinedBoundingBox(axisAlignedBB, decode.getRed(), decode.getGreen(), decode.getBlue(), decode.getAlpha());
+        }
+
+        public static void drawOutlinedBoundingBox(AxisAlignedBB boundingBox, int red, int green, int blue, int alpha)
+        {
+            Tessellator tessellator = Tessellator.getInstance();
+            BufferBuilder vertexbuffer = tessellator.getBuffer();
+            vertexbuffer.begin(3, DefaultVertexFormats.POSITION_COLOR);
+            vertexbuffer.pos(boundingBox.minX, boundingBox.minY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.maxX, boundingBox.minY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.maxX, boundingBox.minY, boundingBox.maxZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.minX, boundingBox.minY, boundingBox.maxZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.minX, boundingBox.minY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            tessellator.draw();
+            vertexbuffer.begin(3, DefaultVertexFormats.POSITION_COLOR);
+            vertexbuffer.pos(boundingBox.minX, boundingBox.maxY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.maxX, boundingBox.maxY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.maxX, boundingBox.maxY, boundingBox.maxZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.minX, boundingBox.maxY, boundingBox.maxZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.minX, boundingBox.maxY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            tessellator.draw();
+            vertexbuffer.begin(1, DefaultVertexFormats.POSITION_COLOR);
+            vertexbuffer.pos(boundingBox.minX, boundingBox.minY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.minX, boundingBox.maxY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.maxX, boundingBox.minY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.maxX, boundingBox.maxY, boundingBox.minZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.maxX, boundingBox.minY, boundingBox.maxZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.maxX, boundingBox.maxY, boundingBox.maxZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.minX, boundingBox.minY, boundingBox.maxZ).color(red, green, blue, alpha).endVertex();
+            vertexbuffer.pos(boundingBox.minX, boundingBox.maxY, boundingBox.maxZ).color(red, green, blue, alpha).endVertex();
+            tessellator.draw();
+        }
+    //
 }
Index: src/minecraft/net/minecraft/client/renderer/chunk/RenderChunk.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/renderer/chunk/RenderChunk.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/renderer/chunk/RenderChunk.java	(date 1502999984000)
@@ -1,6 +1,9 @@
 package net.minecraft.client.renderer.chunk;
 
 import com.google.common.collect.Sets;
+
+import de.paxii.clarinet.module.render.ModuleXray;
+
 import java.nio.FloatBuffer;
 import java.util.HashSet;
 import java.util.Set;
@@ -300,7 +303,13 @@
                 chunkcacheof.renderFinish();
             }
 
-            compiledchunk.setVisibility(lvt_9_1_.computeVisibility());
+            // TODO: Client
+                SetVisibility visibility = lvt_9_1_.computeVisibility();
+                if (ModuleXray.isActive()) {
+                    visibility.setAllVisible(true);
+                }
+                compiledchunk.setVisibility(visibility);
+            //
             this.lockCompileTask.lock();
 
             try
Index: src/minecraft/net/minecraft/client/renderer/entity/Render.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/renderer/entity/Render.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/renderer/entity/Render.java	(date 1502999984000)
@@ -1,5 +1,7 @@
 package net.minecraft.client.renderer.entity;
 
+import de.paxii.clarinet.module.render.ModuleNameTags;
+
 import javax.annotation.Nullable;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.client.Minecraft;
@@ -386,7 +388,14 @@
             boolean flag1 = this.renderManager.options.thirdPersonView == 2;
             float f2 = entityIn.height + 0.5F - (flag ? 0.25F : 0.0F);
             int i = "deadmau5".equals(str) ? -10 : 0;
-            EntityRenderer.drawNameplate(this.getFontRendererFromRenderManager(), str, (float)x, (float)y + f2, (float)z, i, f, f1, flag1, flag);
+
+            // TODO: Client
+                if (ModuleNameTags.isActive()) {
+                    ModuleNameTags.drawHealthTags(entityIn, this.getFontRendererFromRenderManager(), str, (float)x, (float)y + f2, (float)z, i, f, f1, flag1, flag);
+                } else {
+                    EntityRenderer.drawNameplate(this.getFontRendererFromRenderManager(), str, (float)x, (float)y + f2, (float)z, i, f, f1, flag1, flag);
+                }
+            //
         }
     }
 
Index: src/minecraft/net/minecraft/client/renderer/entity/RenderManager.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/renderer/entity/RenderManager.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/renderer/entity/RenderManager.java	(date 1502999984000)
@@ -1,6 +1,11 @@
 package net.minecraft.client.renderer.entity;
 
 import com.google.common.collect.Maps;
+
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.entity.PostRenderEntityEvent;
+import de.paxii.clarinet.event.events.entity.PreRenderEntityEvent;
+
 import java.util.Collections;
 import java.util.Map;
 import javax.annotation.Nullable;
@@ -391,6 +396,16 @@
 
     public void doRenderEntity(Entity entityIn, double x, double y, double z, float yaw, float partialTicks, boolean p_188391_10_)
     {
+        //TODO: Client
+            PreRenderEntityEvent preRenderEvent = new PreRenderEntityEvent(entityIn, x, y, z, partialTicks);
+
+            EventManager.call(preRenderEvent);
+
+            if (preRenderEvent.isCancelled()) {
+                return;
+            }
+        //
+
         Render<Entity> render = null;
 
         try
@@ -441,8 +456,7 @@
                 }
             }
         }
-        catch (Throwable throwable3)
-        {
+        catch (Throwable throwable3) {
             CrashReport crashreport = CrashReport.makeCrashReport(throwable3, "Rendering entity in world");
             CrashReportCategory crashreportcategory = crashreport.makeCategory("Entity being rendered");
             entityIn.addEntityCrashInfo(crashreportcategory);
@@ -453,6 +467,12 @@
             crashreportcategory1.addCrashSection("Delta", Float.valueOf(partialTicks));
             throw new ReportedException(crashreport);
         }
+        //TODO: Client
+            finally {
+                PostRenderEntityEvent postRenderEvent = new PostRenderEntityEvent(entityIn, x, y, z, partialTicks);
+                EventManager.call(postRenderEvent);
+            }
+        //
     }
 
     public void renderMultipass(Entity p_188389_1_, float p_188389_2_)
Index: src/minecraft/net/minecraft/client/settings/KeyBinding.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/client/settings/KeyBinding.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/client/settings/KeyBinding.java	(date 1502999984000)
@@ -2,6 +2,10 @@
 
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
+
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.game.KeyPressedEvent;
+
 import java.util.Map;
 import java.util.Set;
 import java.util.function.Supplier;
@@ -28,6 +32,10 @@
     {
         if (keyCode != 0)
         {
+            //TODO: Client
+                EventManager.call(new KeyPressedEvent(keyCode));
+            //
+
             KeyBinding keybinding = HASH.lookup(keyCode);
 
             if (keybinding != null)
Index: src/minecraft/net/minecraft/enchantment/EnchantmentHelper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/enchantment/EnchantmentHelper.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/enchantment/EnchantmentHelper.java	(date 1502999984000)
@@ -54,6 +54,13 @@
             {
                 NBTTagCompound nbttagcompound = nbttaglist.getCompoundTagAt(i);
                 Enchantment enchantment = Enchantment.getEnchantmentByID(nbttagcompound.getShort("id"));
+
+                //TODO: Client
+                    if (enchantment == null) {
+                        continue;
+                    }
+                //
+
                 int j = nbttagcompound.getShort("lvl");
 
                 if (enchantment == enchID)
Index: src/minecraft/net/minecraft/entity/Entity.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/entity/Entity.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/entity/Entity.java	(date 1502999984000)
@@ -3,6 +3,12 @@
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
+
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.entity.EntityMoveEvent;
+import de.paxii.clarinet.event.events.entity.PostEntityUpdateEvent;
+import de.paxii.clarinet.event.events.player.PlayerStepEvent;
+
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
@@ -22,6 +28,7 @@
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.block.state.pattern.BlockPattern;
+import net.minecraft.client.entity.EntityPlayerSP;
 import net.minecraft.command.CommandResultStats;
 import net.minecraft.command.ICommandSender;
 import net.minecraft.crash.CrashReport;
@@ -633,6 +640,10 @@
 
         this.firstUpdate = false;
         this.world.profiler.endSection();
+
+        //TODO: Client
+            EventManager.call(new PostEntityUpdateEvent(this));
+        //
     }
 
     /**
@@ -722,6 +733,16 @@
      */
     public void move(MoverType type, double x, double y, double z)
     {
+        //TODO: Client
+            EntityMoveEvent moveEvent = new EntityMoveEvent(this, x, y, z, this.noClip);
+            EventManager.call(moveEvent);
+
+            x = moveEvent.getX();
+            y = moveEvent.getY();
+            z = moveEvent.getZ();
+            this.noClip = moveEvent.isNoClip();
+        //
+
         if (this.noClip)
         {
             this.setEntityBoundingBox(this.getEntityBoundingBox().offset(x, y, z));
@@ -917,6 +938,10 @@
 
             if (this.stepHeight > 0.0F && flag && (d2 != x || d4 != z))
             {
+                //TODO: Client
+                    double oldY = this.boundingBox.minY;
+                //
+
                 double d14 = x;
                 double d6 = y;
                 double d7 = z;
@@ -1015,6 +1040,12 @@
                     z = d7;
                     this.setEntityBoundingBox(axisalignedbb1);
                 }
+
+                //TODO: Client
+                    if (this instanceof EntityPlayerSP && oldY < this.boundingBox.minY) {
+                        EventManager.call(new PlayerStepEvent(this.boundingBox.minY - oldY));
+                    }
+                //
             }
 
             this.world.profiler.endSection();
Index: src/minecraft/net/minecraft/entity/EntityLivingBase.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/entity/EntityLivingBase.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/entity/EntityLivingBase.java	(date 1502999984000)
@@ -2,6 +2,9 @@
 
 import com.google.common.base.Objects;
 import com.google.common.collect.Maps;
+
+import de.paxii.clarinet.module.movement.ModuleStrafe;
+
 import java.util.Collection;
 import java.util.ConcurrentModificationException;
 import java.util.Iterator;
@@ -2120,30 +2123,32 @@
                         float f6 = 0.91F;
                         BlockPos.PooledMutableBlockPos blockpos$pooledmutableblockpos = BlockPos.PooledMutableBlockPos.retain(this.posX, this.getEntityBoundingBox().minY - 1.0D, this.posZ);
 
-                        if (this.onGround)
-                        {
-                            f6 = this.world.getBlockState(blockpos$pooledmutableblockpos).getBlock().slipperiness * 0.91F;
-                        }
+                        // TODO: Client
+                            if (this.onGround || ModuleStrafe.isActive())
+                            {
+                                f6 = this.world.getBlockState(blockpos$pooledmutableblockpos).getBlock().slipperiness * 0.91F;
+                            }
 
-                        float f7 = 0.16277136F / (f6 * f6 * f6);
-                        float f8;
+                            float f7 = 0.16277136F / (f6 * f6 * f6);
+                            float f8;
 
-                        if (this.onGround)
-                        {
-                            f8 = this.getAIMoveSpeed() * f7;
-                        }
-                        else
-                        {
-                            f8 = this.jumpMovementFactor;
-                        }
+                            if (this.onGround || ModuleStrafe.isActive())
+                            {
+                                f8 = this.getAIMoveSpeed() * f7;
+                            }
+                            else
+                            {
+                                f8 = this.jumpMovementFactor;
+                            }
 
-                        this.moveRelative(p_191986_1_, p_191986_2_, p_191986_3_, f8);
-                        f6 = 0.91F;
+                            this.moveRelative(p_191986_1_, p_191986_2_, p_191986_3_, f8);
+                            f6 = 0.91F;
 
-                        if (this.onGround)
-                        {
-                            f6 = this.world.getBlockState(blockpos$pooledmutableblockpos.setPos(this.posX, this.getEntityBoundingBox().minY - 1.0D, this.posZ)).getBlock().slipperiness * 0.91F;
-                        }
+                            if (this.onGround || ModuleStrafe.isActive())
+                            {
+                                f6 = this.world.getBlockState(blockpos$pooledmutableblockpos.setPos(this.posX, this.getEntityBoundingBox().minY - 1.0D, this.posZ)).getBlock().slipperiness * 0.91F;
+                            }
+                        //
 
                         if (this.isOnLadder())
                         {
@@ -3187,4 +3192,10 @@
     public void setPartying(BlockPos pos, boolean p_191987_2_)
     {
     }
+
+    //TODO: Client
+        public boolean isUsingItem() {
+            return this.activeItemStack != null;
+        }
+    //
 }
Index: src/minecraft/net/minecraft/entity/projectile/EntityFishHook.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/entity/projectile/EntityFishHook.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/entity/projectile/EntityFishHook.java	(date 1502999984000)
@@ -31,12 +31,17 @@
 import net.minecraft.world.storage.loot.LootContext;
 import net.minecraft.world.storage.loot.LootTableList;
 
+import lombok.Getter;
+
 public class EntityFishHook extends Entity
 {
     private static final DataParameter<Integer> DATA_HOOKED_ENTITY = EntityDataManager.<Integer>createKey(EntityFishHook.class, DataSerializers.VARINT);
     private boolean inGround;
     private int ticksInGround;
-    private EntityPlayer angler;
+    //TODO: Client
+        @Getter
+        private EntityPlayer angler;
+    //
     private int ticksInAir;
     private int ticksCatchable;
     private int ticksCaughtDelay;
Index: src/minecraft/net/minecraft/item/ItemStack.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/item/ItemStack.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/item/ItemStack.java	(date 1502999984000)
@@ -48,12 +48,17 @@
 import net.minecraft.util.text.translation.I18n;
 import net.minecraft.world.World;
 
+import lombok.Getter;
+
 public final class ItemStack
 {
     public static final ItemStack EMPTY = new ItemStack((Item)null);
     public static final DecimalFormat DECIMALFORMAT = new DecimalFormat("#.##");
 
     /** Size of the stack. */
+    //TODO: Client
+        @Getter
+    //
     private int stackSize;
 
     /**
Index: src/minecraft/net/minecraft/network/handshake/client/C00Handshake.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/network/handshake/client/C00Handshake.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/network/handshake/client/C00Handshake.java	(date 1502999984000)
@@ -1,5 +1,7 @@
 package net.minecraft.network.handshake.client;
 
+import de.paxii.clarinet.util.protocol.ProtocolVersion;
+
 import java.io.IOException;
 import net.minecraft.network.EnumConnectionState;
 import net.minecraft.network.Packet;
@@ -19,7 +21,9 @@
 
     public C00Handshake(String p_i47613_1_, int p_i47613_2_, EnumConnectionState p_i47613_3_)
     {
-        this.protocolVersion = 338;
+        // TODO: Client
+            this.protocolVersion = ProtocolVersion.getProtocolVersion();
+        //
         this.ip = p_i47613_1_;
         this.port = p_i47613_2_;
         this.requestedState = p_i47613_3_;
Index: src/minecraft/net/minecraft/network/play/client/CPacketPlayer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/network/play/client/CPacketPlayer.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/network/play/client/CPacketPlayer.java	(date 1502999984000)
@@ -5,14 +5,24 @@
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.INetHandlerPlayServer;
 
+import lombok.Getter;
+import lombok.Setter;
+
 public class CPacketPlayer implements Packet<INetHandlerPlayServer>
 {
     protected double x;
-    protected double y;
+    //TODO: Client
+        @Getter
+        @Setter
+        protected double y;
+    //
     protected double z;
     protected float yaw;
     protected float pitch;
-    protected boolean onGround;
+    //TODO: Client
+        @Setter
+        protected boolean onGround;
+    //
     protected boolean moving;
     protected boolean rotating;
 
Index: src/minecraft/net/minecraft/network/play/server/SPacketEntityVelocity.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/network/play/server/SPacketEntityVelocity.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/network/play/server/SPacketEntityVelocity.java	(date 1502999984000)
@@ -6,12 +6,19 @@
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.INetHandlerPlayClient;
 
+import lombok.Setter;
+
 public class SPacketEntityVelocity implements Packet<INetHandlerPlayClient>
 {
     private int entityID;
-    private int motionX;
-    private int motionY;
-    private int motionZ;
+    //TODO: Client
+        @Setter
+        private int motionX;
+        @Setter
+        private int motionY;
+        @Setter
+        private int motionZ;
+    //
 
     public SPacketEntityVelocity()
     {
Index: src/minecraft/net/minecraft/realms/RealmsSharedConstants.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/realms/RealmsSharedConstants.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/realms/RealmsSharedConstants.java	(date 1502999984000)
@@ -1,11 +1,15 @@
 package net.minecraft.realms;
 
+import de.paxii.clarinet.util.protocol.ProtocolVersion;
+
 import net.minecraft.util.ChatAllowedCharacters;
 
 public class RealmsSharedConstants
 {
-    public static int NETWORK_PROTOCOL_VERSION = 338;
-    public static int TICKS_PER_SECOND = 20;
-    public static String VERSION_STRING = "1.12.1";
-    public static char[] ILLEGAL_FILE_CHARACTERS = ChatAllowedCharacters.ILLEGAL_FILE_CHARACTERS;
+    // TODO: Client
+        public static int NETWORK_PROTOCOL_VERSION = ProtocolVersion.getProtocolVersion();
+        public static int TICKS_PER_SECOND = 20;
+        public static String VERSION_STRING = ProtocolVersion.getGameVersion();
+        public static char[] ILLEGAL_FILE_CHARACTERS = ChatAllowedCharacters.ILLEGAL_FILE_CHARACTERS;
+    //
 }
Index: src/minecraft/net/minecraft/server/MinecraftServer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/server/MinecraftServer.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/server/MinecraftServer.java	(date 1502999984000)
@@ -9,6 +9,9 @@
 import com.mojang.authlib.GameProfileRepository;
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
+
+import de.paxii.clarinet.util.protocol.ProtocolVersion;
+
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufOutputStream;
 import io.netty.buffer.Unpooled;
@@ -529,7 +532,9 @@
                 this.currentTime = getCurrentTimeMillis();
                 long i = 0L;
                 this.statusResponse.setServerDescription(new TextComponentString(this.motd));
-                this.statusResponse.setVersion(new ServerStatusResponse.Version("1.12.1", 338));
+                // TODO: Client
+                    this.statusResponse.setVersion(new ServerStatusResponse.Version(ProtocolVersion.getGameVersion(), ProtocolVersion.getProtocolVersion()));
+                //
                 this.applyServerIconToResponse(this.statusResponse);
 
                 while (this.serverRunning)
Index: src/minecraft/net/minecraft/server/network/NetHandlerHandshakeTCP.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/server/network/NetHandlerHandshakeTCP.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/server/network/NetHandlerHandshakeTCP.java	(date 1502999984000)
@@ -1,5 +1,7 @@
 package net.minecraft.server.network;
 
+import de.paxii.clarinet.util.protocol.ProtocolVersion;
+
 import net.minecraft.network.EnumConnectionState;
 import net.minecraft.network.NetworkManager;
 import net.minecraft.network.handshake.INetHandlerHandshakeServer;
@@ -32,13 +34,15 @@
             case LOGIN:
                 this.networkManager.setConnectionState(EnumConnectionState.LOGIN);
 
-                if (packetIn.getProtocolVersion() > 338)
+                // TODO: Client
+                if (packetIn.getProtocolVersion() > ProtocolVersion.getProtocolVersion())
                 {
                     ITextComponent itextcomponent = new TextComponentTranslation("multiplayer.disconnect.outdated_server", new Object[] {"1.12.1"});
                     this.networkManager.sendPacket(new SPacketDisconnect(itextcomponent));
                     this.networkManager.closeChannel(itextcomponent);
                 }
-                else if (packetIn.getProtocolVersion() < 338)
+                // TODO: Client
+                else if (packetIn.getProtocolVersion() < ProtocolVersion.getProtocolVersion())
                 {
                     ITextComponent itextcomponent1 = new TextComponentTranslation("multiplayer.disconnect.outdated_client", new Object[] {"1.12.1"});
                     this.networkManager.sendPacket(new SPacketDisconnect(itextcomponent1));
Index: src/minecraft/net/minecraft/util/MovementInputFromOptions.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/util/MovementInputFromOptions.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/util/MovementInputFromOptions.java	(date 1502999984000)
@@ -1,5 +1,8 @@
 package net.minecraft.util;
 
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.player.UpdatePlayerMoveStateEvent;
+
 import net.minecraft.client.settings.GameSettings;
 
 public class MovementInputFromOptions extends MovementInput
@@ -64,5 +67,11 @@
             this.moveStrafe = (float)((double)this.moveStrafe * 0.3D);
             this.moveForward = (float)((double)this.moveForward * 0.3D);
         }
+
+        //TODO: Client
+            UpdatePlayerMoveStateEvent moveStateEvent = EventManager.call(new UpdatePlayerMoveStateEvent(this.moveStrafe, this.moveForward));
+            this.moveStrafe = moveStateEvent.getMoveStrafe();
+            this.moveForward = moveStateEvent.getMoveForward();
+        //
     }
 }
Index: src/minecraft/net/minecraft/util/math/MathHelper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/util/math/MathHelper.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/util/math/MathHelper.java	(date 1502999984000)
@@ -64,6 +64,12 @@
         return (float)Math.sqrt(value);
     }
 
+    //TODO: Client
+        public static float sqrt_double(double value) {
+            return sqrt(value);
+        }
+    //
+
     /**
      * Returns the greatest integer less than or equal to the float argument
      */
Index: src/minecraft/net/minecraft/util/text/TextFormatting.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/util/text/TextFormatting.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/util/text/TextFormatting.java	(date 1502999984000)
@@ -9,6 +9,8 @@
 import java.util.regex.Pattern;
 import javax.annotation.Nullable;
 
+import lombok.Getter;
+
 public enum TextFormatting
 {
     BLACK("BLACK", '0', 0),
@@ -46,7 +48,10 @@
     private final String name;
 
     /** The formatting code that produces this format. */
-    private final char formattingCode;
+    //TODO: Client
+        @Getter
+        private final char formattingCode;
+    //
     private final boolean fancyStyling;
 
     /**
Index: src/minecraft/net/minecraft/world/World.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/world/World.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/world/World.java	(date 1502999984000)
@@ -4,6 +4,11 @@
 import com.google.common.base.MoreObjects;
 import com.google.common.base.Predicate;
 import com.google.common.collect.Lists;
+
+import de.paxii.clarinet.Wrapper;
+import de.paxii.clarinet.event.EventManager;
+import de.paxii.clarinet.event.events.game.LoadWorldEvent;
+
 import java.util.Calendar;
 import java.util.Collection;
 import java.util.Iterator;
@@ -1236,6 +1241,12 @@
      */
     public boolean spawnEntity(Entity entityIn)
     {
+        //TODO: Client
+            if (entityIn == Wrapper.getPlayer()) {
+                EventManager.call(new LoadWorldEvent(this));
+            }
+        //
+
         int i = MathHelper.floor(entityIn.posX / 16.0D);
         int j = MathHelper.floor(entityIn.posZ / 16.0D);
         boolean flag = entityIn.forceSpawn;
@@ -4011,4 +4022,66 @@
     {
         return null;
     }
+
+    //TODO: Client
+        public Block getBlock(int x, int y, int z) {
+            BlockPos blockPos = new BlockPos(x, y, z);
+
+            return this.getBlockState(blockPos).getBlock();
+        }
+
+        public List<AxisAlignedBB> getCollidingBoundingBoxes(Entity entityIn, AxisAlignedBB bb) {
+            return this.getCollisionBoxes(entityIn, bb);
+        }
+
+        public List<AxisAlignedBB> isAABBInMaterial(@Nullable Entity entityIn, AxisAlignedBB aabb) {
+            return this.getCollisionBoxes(entityIn, aabb);
+        }
+
+        /**
+         * checks if the given AABB is in the material given. Used while swimming.
+         */
+        public boolean isAABBInMaterial(AxisAlignedBB bb, Material materialIn)
+        {
+            int i = MathHelper.floor(bb.minX);
+            int j = MathHelper.ceil(bb.maxX);
+            int k = MathHelper.floor(bb.minY);
+            int l = MathHelper.ceil(bb.maxY);
+            int i1 = MathHelper.floor(bb.minZ);
+            int j1 = MathHelper.ceil(bb.maxZ);
+            BlockPos.PooledMutableBlockPos blockpos$pooledmutableblockpos = BlockPos.PooledMutableBlockPos.retain();
+
+            for (int k1 = i; k1 < j; ++k1)
+            {
+                for (int l1 = k; l1 < l; ++l1)
+                {
+                    for (int i2 = i1; i2 < j1; ++i2)
+                    {
+                        IBlockState iblockstate = this.getBlockState(blockpos$pooledmutableblockpos.setPos(k1, l1, i2));
+
+                        if (iblockstate.getMaterial() == materialIn)
+                        {
+                            int j2 = (iblockstate.getValue(BlockLiquid.LEVEL)).intValue();
+                            double d0 = (double)(l1 + 1);
+
+                            if (j2 < 8)
+                            {
+                                d0 = (double)(l1 + 1) - (double)j2 / 8.0D;
+                            }
+
+                            if (d0 >= bb.minY)
+                            {
+                                blockpos$pooledmutableblockpos.release();
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+
+            blockpos$pooledmutableblockpos.release();
+            return false;
+        }
+    //
+
 }
Index: src/minecraft/net/minecraft/world/WorldProvider.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/world/WorldProvider.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/world/WorldProvider.java	(date 1502999984000)
@@ -64,7 +64,9 @@
     /**
      * Creates the light to brightness table
      */
-    protected void generateLightBrightnessTable()
+    // TODO: Client
+    public void generateLightBrightnessTable()
+    //
     {
         float f = 0.0F;
 
Index: src/minecraft/net/minecraft/world/WorldProviderHell.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/minecraft/net/minecraft/world/WorldProviderHell.java	(date 1502999936000)
+++ src/minecraft/net/minecraft/world/WorldProviderHell.java	(date 1502999984000)
@@ -33,7 +33,9 @@
     /**
      * Creates the light to brightness table
      */
-    protected void generateLightBrightnessTable()
+    // TODO: Client
+        public void generateLightBrightnessTable()
+    //
     {
         float f = 0.1F;
 
